File: lv2/76502.py
'''
괄호 회전하기
https://school.programmers.co.kr/learn/courses/30/lessons/76502
'''
from collections import deque

def isAvailable(arr):                             # 해당 배열이 올바른 괄호 문자열인지 출력해주는 함수
  stack = []
  pair = {'}' : '{', ']' : '[', ')' : '('}
  for i in arr:
    if len(stack) == 0:                           # stack에 아무것도 없다면 push
      stack.append(i)
    else:
      if i in pair and stack[len(stack)-1] == pair[i]:    # push할 값이 닫는 괄호일 때 앞전 stack 값이 올바른 여는 괄호라면 pop, 아니라면 push
        stack.pop()
      else:
        stack.append(i)
  return True if len(stack) == 0 else False       # stack에 값이 남아있지 않다면 올바른 괄호 문자열

def solution(s):
  answer = 0
  dq = deque(list(s))               # queue와 dequeue의 시간복잡도를 O(1) 로 하기 위해 deque 사용
  for i in range(len(dq)):          
    dq.append(dq.popleft())         # 앞 숫자를 하나 빼서 뒤에 넣는 작업
    if isAvailable(dq):             # 올바른 괄호 문자열이면 정답을 1 추가
      answer += 1
  return answer

File: lv2_js/76502.js
`
괄호 회전하기
https://school.programmers.co.kr/learn/courses/30/lessons/76502
`
function solution(s) {
  let answer = 0;
  const arr = [...s];
  for (let i = 0; i < s.length; i++) {
    arr.unshift(arr.pop()); //arr 에서 하나씩 오른쪽으로 회전
    if (isRight(arr)) {
      answer += 1;
    }
  }
  return answer;
}

function isRight(arr) { //괄호가 올바르게 배치되어 있는지 확인하는 함수
  const mapping = new Map();
  mapping.set(']', '[');
  mapping.set('}', '{');
  mapping.set(')', '(');
  const stack = [];
  for (let i = 0; i < arr.length; i++) {
    if (mapping.has(arr[i])) { // 우괄호 차례인 경우
      if (stack.length > 0 && stack[stack.length - 1] === mapping.get(arr[i])) { // 짝이 맞는 경우
        stack.pop(); //좌괄호를 빼고
      } else {
        return false; //아니라면 짝이 맞지 않으므로 false 
      }
    } else {
      stack.push(arr[i]); // 좌괄호는 무조건 push
    }
  }
  return stack.length === 0 ? true : false; //스택에 값이 남아있으면 괄호가 맞지 않는 것
}